Fixes specifying ruby patch level in gemfile and gemspec

chalkos opened the pull request:
- [x] add failing spec
- [x] add fix (with passing spec) and more specs to cover edge cases
- [x] ~~change error message to a more meaningful one~~ fixed the bug and a meaningful error message is now being shown

spec always passes on 2.2.0p0, probably because the patch version is 0

chalkos commented:
My findings so far:

(On brach master)
In Bundler::Resolver::SpecGroup#for? (bundler/resolver.rb) in `spec_required_ruby_version.satisfied_by?(required_ruby_version.gem_version)`, `spec_required_ruby_version` has a ruby version with the patch number, but `required_ruby_version.gem_version` only contains the version without patch number (is it present in `required_ruby_version.patch_version`).

@segiddins do you think `required_ruby_version.gem_version` should include the patch number? I can track down where it is being set and make sure the patch number gets included. I believe this would be the best way to fix this, but as it may also have unintended effects on other parts of the code I could also change `Bundler::Resolver::SpecGroup#for?` to use a patch version when it is present.

The `Gem::Version` assumes a patch number of 0, that is why the tests for `ruby 2.2.0p0` pass: because `Gem::Requirement#satisfied_by?` returns true for that case.

segiddins commented:
@chalkos let's just see if changing `gem_version` fixes things?

segiddins commented on lib/bundler/resolver.rb:
```diff
@@ -131,7 +131,7 @@ def source
         if spec = @specs[platform]
           if required_ruby_version && spec.respond_to?(:required_ruby_version) && spec_required_ruby_version = spec.required_ruby_version
-            spec_required_ruby_version.satisfied_by?(required_ruby_version.gem_version)
+            spec_required_ruby_version.satisfied_by?(Gem::Version.new("#{required_ruby_version.gem_version}.#{required_ruby_version.patchlevel}"))
```
`required_ruby_version.patchlevel` can be nil, and this should be a method on `Bundler::RubyVersion`

chalkos replied:
Thanks. I tried changing `Bundler::RubyVersion#initialize` to always include the patch version in `@gem_version`, but some tests started failing. From your experience, do you think it would be better to do this or handle this specific case in `#for?` ?

Maybe in some parts of bundler it is preferable not to have the patch version?

segiddins replied:
Maybe just a `gem_version_with_patch`?

chalkos replied:
That seems cleaner and reusable. I'll do that.

chalkos commented:
I think this could use some more specs.
The new method in `Bundler::RubyVersion` could (needs to?) be tested.
Regarding #4593, I think some more tests could be written, specially for some edge cases.

segiddins commented on lib/bundler/ruby_version.rb:
```diff
@@ -100,6 +100,19 @@ def self.system
+      unless @gem_version_with_patch
+        @gem_version_with_patch = Gem::Requirement.create("#{@gem_version}.0").requirements.first.last
+        if @patchlevel
+          begin
```
the `begin` is unnecessary

chalkos replied:
it's meant to handle the `BadRequirementError` exception. But I'm not really sure this even works for that case. I found a better way to do this already

segiddins commented on lib/bundler/ruby_version.rb:
```diff
@@ -100,6 +100,19 @@ def self.system
     end
 
+    def gem_version_with_patch
+      unless @gem_version_with_patch
```
`||=`

chalkos replied:
right! 👍

segiddins commented on lib/bundler/ruby_version.rb:
```diff
@@ -100,6 +100,19 @@ def self.system
 
+    def gem_version_with_patch
+      unless @gem_version_with_patch
+        @gem_version_with_patch = Gem::Requirement.create("#{@gem_version}.0").requirements.first.last
```
this unconditionally creates an object that might not be used

chalkos replied:
right. my next commit will change this

segiddins commented:
@chalkos is this still WIP?

chalkos commented:
Yes. I still need to add more specs to cover edge cases and to RubyVersion's new method and change error message to a more meaningful one.

segiddins commented on lib/bundler/ruby_version.rb:
```diff
@@ -100,6 +100,15 @@ def self.system
+    def to_gem_version_with_patchlevel
+      @gem_version_with_patch ||= begin
+        patch_number = @patchlevel ? Gem::Requirement.create(@patchlevel).requirements.first.last : 0
+        Gem::Requirement.create("#{@gem_version}.#{patch_number}").requirements.first.last
```
this can just be `Gem::Version.create("#{@gem_version}.#{patch_number}")`

segiddins commented on lib/bundler/ruby_version.rb:
```diff
@@ -100,6 +100,15 @@ def self.system
+        patch_number = @patchlevel ? Gem::Requirement.create(@patchlevel).requirements.first.last : 0
+        Gem::Requirement.create("#{@gem_version}.#{patch_number}").requirements.first.last
+      rescue BadRequirementError
+        Gem::Requirement.create("#{@gem_version}.0").requirements.first.last
```
this can just be `@gem_version`

chalkos replied:
I wanted to give it a better name than that. `@gem_version` does not seem like "a representation of the RubyVersion as a Gem::Version object", which is what the instance variable represents.

segiddins replied:
No i mean there's no reason to create a new version object in this case

chalkos replied:
Sorry, I misunderstood and thought you were talking about something else.
You are correct, I'll change this to `@gem_version`

chalkos commented:
@segiddins after making the changes you suggested, would you like me to separate this in 2 or 3 commits? I think we could have one with the initial specs and then another one or two with the fix and the RubyVersion specs.

segiddins commented on lib/bundler/ruby_version.rb:
```diff
@@ -100,6 +100,15 @@ def self.system
 
+    def to_gem_version_with_patchlevel
+      @gem_version_with_patch ||= begin
+        patch_number = @patchlevel ? Gem::Requirement.create(@patchlevel).requirements.first.last : 0
```
why can't this just be `@patchlevel` ?

chalkos replied:
I think it's exactly the same. I'll change it, yours is simpler.
The purpose is to avoid raising and catching an exception when `@patchlevel` is nil. Although the exception is still risen and caught if `@patchlevel` is negative or some other invalid value.

I think that performance-wise it better to handle these kind of situations using conditions instead of catching the exception. But if I were to fully validate the `@patchlevel`, it would be like duplicating the code in Gem::Version that already does that.

I think that the code that just catches the exception instead of using conditions would be easier to understand:

```
@gem_version_with_patch ||= begin
  Gem::Version.create("#{@gem_version}.#{@patchlevel}")
rescue BadRequirementError
  Gem::Version.create(@gem_version)
end
```

Which one would you consider to be better?

segiddins replied:
```ruby
@gem_version_with_patch ||= begin
  Gem::Version.create("#{@gem_version}.#{@patchlevel}")
rescue BadRequirementError
  @gem_version
end
```

segiddins commented on spec/install/gemspecs_spec.rb:
```diff
@@ -46,4 +46,76 @@
+
+    it "installs when patch level is specified and the version still matches the current version" do
+      build_lib("foo", :path => tmp.join("foo")) do |s|
+        s.write "Gemfile", <<-G
```
please use `install_gemfile` to install a gemfile, rather than writing it into the lib path

chalkos replied:
I looked for a way to set `Gem::Specification#required_ruby_version` with `install_gemfile`, but I could not find one. Is it possible?

indirect replied:
Setting the required ruby version here is correct, but you can use `install_gemfile` to create and install the Gemfile.

indirect replied:
I would suggest something like this:

```ruby
build_lib("foo", :path => bundled_app) do |s|
  s.required_ruby_version = RUBY_VERSION
end

install_gemfile <<-G
  ruby '#{RUBY_VERSION}', :engine_version => '#{RUBY_VERSION}', :engine => 'ruby'
  gemspec
G
should_be_installed "foo 1.0"
```

segiddins commented:
@bundlerbot r+

indirect commented:
woo bundlerbot! :tada:

segiddins commented:
@chalkos see https://travis-ci.org/bundler/bundler/jobs/136049835#L3428-L3446

chalkos commented:
I'll rebase it to reproduce this on my end.

chalkos commented:
@segiddins turns out `ruby 2.4.0dev` has a `RUBY_PATCHLEVEL` of -1
For these versions you can't specify a patchlevel in the gemspec (because it would be -1 and bundler will complain about that). I suggest that affected tests are skipped in such cases.

segiddins commented:
@chalkos agreed

chalkos commented:
I rebased on master and added code that should skip the spec on ruby versions with negative patchlevel.
@segiddins is this supposed to fail? https://travis-ci.org/bundler/bundler/jobs/136242613

segiddins commented:
@homu r+

homu commented:
:pushpin: Commit 4c5d334 has been approved by `segiddins`

<!-- @homu r=segiddins 4c5d334bea9431ba51d3095e0d4fa16804d86268 -->

homu commented:
:hourglass: Testing commit 4c5d334 with merge 64edbb8...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/bundler/bundler/builds/136296915)
